Space time profile

spencerjolly opened the pull request:
This adds the ability to create a pulse with space-time couplings. Spatial chirp at the focus (z=0) can be added with a real-valued parameter 'sc'. This uses the highest level 'Profile' class. An interesting test is if the propagator works properly to propagate such pulses to the beginning of a simulation box. Analytical equations could also be used to directly calculate the fields at a user-defined z-position.

spencerjolly committed:
Add files via upload

space-time profile

spencerjolly committed:
Update spacetime_profile.py

spencerjolly committed:
Update spacetime_profile.py

spencerjolly committed:
Update spacetime_profile.py

spencerjolly committed:
correct time variable name

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

spencerjolly committed:
fix typos and remove mention of pft

spencerjolly committed:
style changes

spencerjolly committed:
remove spurious indent

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

RemiLehe commented on lasy/profiles/spacetime_profile.py:
```diff
@@ -0,0 +1,102 @@
+from .profile import Profile
+
+
+class SpaceTimeProfile(Profile):
```
Maybe this could actually be part of the `GaussianProfile` laser. See other comments.

spencerjolly replied:
I don't actually see how it could be applied with GaussianProfile. In that class it calls the temporal and spatial profiles separately, but we need access to the space and time dimensions at the same time.

RemiLehe replied:
Yes, agreed. If we decide to have it applied with GaussianProfile, we would completely change the implementation of the `GaussianProfile` (which is definitely possible): instead of calling the temporal and spatial profile separately, we would have the Python code for the gaussian profile directly in the body of the `GaussianProfile` class. Does that work for you?

spencerjolly replied:
Yes, ok with me. In fact, it would mostly entail pasting the code from this PR to the GaussianProfile class. However, I also wanted to add a pulse-front tilt parameter (in addition to the current "spatial chirp" parameter). I could do that in this PR so that it is all there, and then we could paste it.

Let me know how you want to continue.

RemiLehe replied:
Sounds good. I think it is good to do all of the above as part of the current PR, i.e. adding spatial chirp, and moving all of this to the `GaussianProfile` class.

spencerjolly replied:
I've done these changes. There are some weird test and docs errors at the moment, but in any case I need to build a stand-alone test for the space-time cases. I will try to do it next week.

RemiLehe commented on lasy/profiles/spacetime_profile.py:
```diff
@@ -0,0 +1,102 @@
+    w0: float (in meter)
+        The waist of the laser pulse, i.e. :math:`w_0` in the above formula.
+
+    sc: spatial chirp, b in the above formula, that results in a mixing
+        of the longitudinal and transverse profiles. Must be in units
+        of [x/omega]. A representative real value is b = w0 * tau.
```
Here is a suggestion that is more consistent with the other docstrings:
```suggestion
    b: float (in meter.second)
        Spatial chirp, i.e. :math:`b` in the above formula, that results in the laser frequency
        varying as a function of `x`. A representative real value is b = w0 * tau.

```

RemiLehe reviewed (changes requested):
Thanks for this PR!

I added a few inline comments. In addition, I have a few high-level comments:
- Would it be OK to include the parameter `b` in the [`GaussianProfile`](https://lasydoc.readthedocs.io/en/latest/api/profiles/gaussian.html) and have its value be 0 by default, instead of having a separate `SpaceTimeProfile` class? (This is the approach that we adopted for the built-it Gaussian profile in WarpX: https://github.com/ECP-WarpX/WarpX/blob/development/Source/Laser/LaserProfilesImpl/LaserProfileGaussian.cpp#L99 ; it would require reimplementing the `GaussianProfile` differently in WarpX.)
- Could you add an automated test, e.g. [here](https://github.com/LASY-org/lasy/blob/development/tests/test_laser_profiles.py)? The test could for instance check for instance that the laser duration is unchanged in the presence of spatial chirp, or (even better) check that the laser frequency is indeed `x` dependent.

spencerjolly committed:
Update lasy/profiles/spacetime_profile.py

update docstring

Co-authored-by: Remi Lehe <remi.lehe@normalesup.org>

spencerjolly committed:
change sc to b in spacetime

spencerjolly committed:
change sc to b in spacetime

spencerjolly committed:
Update gaussian_profile.py

Move space-time effects to Gaussian profile

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

spencerjolly committed:
Update gaussian_profile.py

Dependencies fix

spencerjolly committed:
Update test_gaussian_propagator.py

Update GaussianProfile call

spencerjolly committed:
Update test_laser_profiles.py

Update GaussianProfile call

spencerjolly committed:
Update test_laser_profiles.py

re-order GaussianProfile arguments

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

spencerjolly committed:
Update test_gaussian_propagator.py

re-order GaussianProfile arguments

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

spencerjolly committed:
Update gaussian_profile.py

re-order space-time arguments

spencerjolly committed:
Update test_gaussian_propagator.py

remove trans_profile mention

spencerjolly committed:
Update test_laser_utils.py

remove long_profile mention

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

spencerjolly committed:
Update test_t2z2t.py

remove trans_profile and long_profile mentions

spencerjolly committed:
Update test_t2z2t.py

remove trans_profile and long_profile mentions

spencerjolly committed:
Update test_gaussian_propagator.py

spencerjolly committed:
Update gaussian_profile.py

spencerjolly committed:
Update laser_utils.py

New function get_t_peak that calculates the average time of the intensity. Relevant for pulse-front tilt testing.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

spencerjolly committed:
Update laser_utils.py

spencerjolly committed:
Update test_laser_utils.py

Add t_peak test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

spencerjolly committed:
Update test_laser_utils.py

spencerjolly committed:
Update test_laser_utils.py

spencerjolly committed:
Update test_laser_utils.py

spencerjolly committed:
Update test_gaussian_propagator.py

spencerjolly committed:
Update test_gaussian_propagator.py

spencerjolly committed:
Update test_gaussian_propagator.py

spencerjolly committed:
Update test_gaussian_propagator.py

spencerjolly committed:
Update test_laser_utils.py

spencerjolly committed:
Update test_laser_profiles.py

spencerjolly committed:
Update test_laser_profiles.py

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

spencerjolly committed:
Update test_laser_profiles.py

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

spencerjolly committed:
Update test_laser_utils.py

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

spencerjolly commented:
@RemiLehe I am slowly progressing on this. Right now I am finishing the tests for angular dispersion and spatial chirp. How can I select the field only at a given transverse position? Basically I want to test either the central frequency or t_peak given spatial chirp or angular dispersion, and for that I would like to have the temporal envelop at y=0 and x=w0.

spencerjolly committed:
Update test_laser_utils.py

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

spencerjolly committed:
Update test_laser_utils.py

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

spencerjolly committed:
Update test_laser_utils.py

MaxThevenet commented:
Thanks for this PR! I think PR #307 implements these changes. @spencerjolly could you check if this works for you?

spencerjolly commented:
@MaxThevenet seems good to me. Really nice to see that it could be implemented together with the tests.
